pkg/healthv2/provider.go

```diff
@@ -0,0 +1,219 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package healthv2
+
+import (
+	"context"
+	"fmt"
+	"sync/atomic"
+	"time"
+
+	"github.com/cilium/cilium/pkg/healthv2/types"
+	"github.com/cilium/cilium/pkg/hive/cell"
+	"github.com/cilium/cilium/pkg/statedb"
+	"github.com/cilium/cilium/pkg/statedb/index"
+	"github.com/sirupsen/logrus"
+)
+
+type HealthProviderParams struct {
+	cell.In
+	Lifecycle cell.Lifecycle
+	DB        *statedb.DB
+}
+
+type HealthProvider struct {
+	db           *statedb.DB
+	stopped      atomic.Bool
+	statusTable  statedb.RWTable[types.Status]
+	Statuses     statedb.Table[types.Status]
+	PrimaryIndex statedb.Index[types.Status, types.HealthID]
+	LevelIndex   statedb.Index[types.Status, types.Level]
+}
+
+const HealthTable = "health"
+
+func newHealthV2Provider(params HealthProviderParams) (types.Provider, *HealthProvider, error) {
+	p := &HealthProvider{
+		db: params.DB,
+	}
+	p.PrimaryIndex = statedb.Index[types.Status, types.HealthID]{
+		Name: "identifier",
+		FromObject: func(s types.Status) index.KeySet {
+			return index.NewKeySet([]byte(s.ID.String()))
+		},
+		FromKey: func(k types.HealthID) index.Key {
```

Comment:
Here as well we can do `index.Stringer[types.HealthID]` (and perhaps Go will be able to infer the type in the future...)